splitio/producer/storage/segment_wrapper.go

```diff
@@ -0,0 +1,141 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/splitio/go-split-commons/v4/storage"
+	"github.com/splitio/go-toolkit/v5/datastructures/set"
+	"github.com/splitio/go-toolkit/v5/logging"
+)
+
+// ObservableSegmentStorage builds on top of the SegmentStorage interface adding some observability methods
+type ObservableSegmentStorage interface {
+	storage.SegmentStorage
+	NamesAndCount() map[string]int
+}
+
+// ObservableSegmentStorageImpl is an implementation of the ObservableSegmentStorage interface
+type ObservableSegmentStorageImpl struct {
+	storage.SegmentStorage
+	counter *activeSegmentTracker
+	logger  logging.LoggerInterface
+}
+
+// NewObservableSegmentStorage constructs and observable segment storage
+func NewObservableSegmentStorage(
+	logger logging.LoggerInterface,
+	splitStorage storage.SplitStorage,
+	toWrap storage.SegmentStorage,
+) *ObservableSegmentStorageImpl {
+
+	if _, ok := toWrap.(supportsUpdateWithSummary); !ok {
+		logger.Warning("supplied segment storage doesn't report added/removed counts, this may introduce inconcistencies in observability endpoints")
```

Comment:
What are your thoughts on not even allowing `ObservableSegmentStorage` to receive a `SegmentStorage` that does not support reporting? Which would allow other functions here to not even check for `supportsUpdateWithSummary` and dealing with both possibilities.

Reply:
Yes, you have a valid point. I recall some constructors returning type-erased interfaces, which means at some point i'll have to do the type-assertion anyways, but I may as well do it earlier enough and avoid those checks. I also don't want to make this interfaces public (since they're subject to change soon).

Would it be ok if I accept the basic interface type (like now), and return an error if it doesn't have the required methods? This would remove the need for those checks every time `Update()` is called.

Reply:
I think that would work well 👍

Reply:
thank you